### PR TITLE
Explore: Use rich history local storage for autocomplete

### LINKED ISSUE #62757: Query History: If not using remote storage, use local storage from autocomplete history
## Background
To enable future growth of query history, we moved query storage from local storage, which exists only on the users computer, to remote storage, which saves queries on the Grafana instance's database. 

### Code change timeline
* Introduced remote storage for query history in #44327
* Enabled by default in #49407
* Migrated local storage to remote storage in #48572
* Migration was removed in https://github.com/grafana/grafana/pull/67470

Initially, we had thought to completely deprecate storage of queries in local storage, but due to another existing feature, we should be able to leave the option in without much extra code. There are also benefits to local storage that we would not want to eliminate completely.

**Remote storage benefits**
- Persists between browsers / computers / after uses clears browser data 
- Can be shareable

**Local storage benefits**
- Usable when Grafana is stateless (the database is not persisted)
- If users at the “viewer” level can access Explore, and the people managing the instance do not want viewers to potentially add much to the database, this would give them that option
     - Potential example use case: a Grafana instance where the user base includes a large number of contracted workers that are not managed directly by the Grafana maintainers. They may be given “viewer with explore” access to keep costs/bandwidth for Grafana storage down. If a large number of users are creating a large number of entries in the database, the instance managers may wish to disable this feature. Local storage would ensure those users still have some amount of persisted query storage on their local machines.

According to our usage data, there are 0 users that are using local storage (data can be provided internally). This shouldn’t be extrapolated to mean absolutely no users use local storage, but it does mean we should prioritize it fairly low.

## Requirements

Change the autocomplete feature ([src](https://github.com/grafana/grafana/blob/main/public/app/core/utils/explore.ts#L298)) to use the query history local storage. Saving to local storage should happen regardless if remote storage is being used or not.  Reading query history should depend on the setting, and reading autocomplete should always read from local storage. Writing to remote storage should only happen if the setting is enabled.

Autocomplete does not have the concept of "starring" queries to save them for a more permanent amount of time. Disable the ability to star queries if local storage is being used, and update the documentation accordingly.

Limit this local storage to only save 100 entries. This is to not overwhelm the autocomplete feature.

Remove existing autocomplete storage functionality. 

Current Query history local storage (read/write) code is at https://github.com/grafana/grafana/blob/main/public/app/core/history/RichHistoryLocalStorage.ts 

Current autocomplete history code (write) is at 
https://github.com/grafana/grafana/blob/main/public/app/core/utils/explore.ts#L279

Ensure current autocomplete use (read) is functional
https://github.com/grafana/grafana/blob/main/public/app/features/explore/state/utils.ts#L122

### PR DESCRIPTION
**Pull request by gelicia**
**What is this feature?**

This removes local storage use for auto complete and gets autocomplete history data from the local storage used for query history.

Query history is now saved always to local storage, and optionally to remote storage if the feature flag is enabled. 

**Why do we need this feature?**

Autocomplete and query history local storage did redundant tasks. This removes the redundancy in local storage. 

**Which issue(s) does this PR fix?**:

Fixes #62757

**Special notes for your reviewer:**

There is a bit of a difference in how mixed datasource queries are handled. For example, with loki, if you run 3 queries with the test datasource and then one in loki, it will end up sending all 4 queries to the datasource as history. It filters out all queries without an expr ([src](https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/CompletionDataProvider.ts#L35)) so only the loki one shows. That said, if another datasource uses the field expr in a query history entry, it will show.

With this, I first query local storage for the selected datasource and pull those entries. If the number of queries is under the max (currently 100), I query the mixed datasource for the max number, and append on any with queries that match the datasource. This means better validation that the datasource is only getting sent relevant history. 

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/features/explore/state/history.ts [hunk 1]
```diff
@@ -74,12 +74,14 @@ const forEachExplorePane = (state: ExploreState, callback: (item: ExploreItemSta
 };
 
 export const addHistoryItem = (
+  localOverride: boolean,
```
[the remark this answers is not in the document]
Comment (gelicia) [reply public/app/features/explore/state/history.ts]: Good idea! fixed!

File: public/app/features/explore/state/query.ts [hunk 2]
```diff
@@ -485,11 +484,10 @@ async function handleHistory(
...
   exploreId: string
 ) {
-  const datasourceId = datasource.meta.id;
-  const nextHistory = updateHistory(history, datasourceId, queries);
-  dispatch(historyUpdatedAction({ exploreId, history: nextHistory }));
-
-  dispatch(addHistoryItem(datasource.uid, datasource.name, queries));
+  dispatch(addHistoryItem(true, datasource.uid, datasource.name, queries)); // always write to local
+  if (config.queryHistoryEnabled) {
+    dispatch(addHistoryItem(false, datasource.uid, datasource.name, queries));
```
[the remark this answers is not in the document]
Comment (gelicia) [reply public/app/features/explore/state/query.ts]: As of how it is now, if the feature flag is off, it won't give any storage full / limit exceeded messages. 

Are you thinking that the write should happen no matter what, but we mute errors if config.queryHistoryEnabled is true, as if so, the write is only happening for autocomplete?

File: public/app/features/explore/state/history.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (gelicia) [reply public/app/features/explore/state/history.ts]: @ifrost actually, it's better we let everything go through richHistory/addToRichHistory and let it call out to the individual storage's addToRichHistory. Reason is the local storage throws a specific error for duplicate queries and doesn't save them (https://github.com/grafana/grafana/blob/main/public/app/core/history/RichHistoryLocalStorage.ts#L70) and the richHistory/addToRichHistory ignores it https://github.com/grafana/grafana/blob/main/public/app/core/utils/richHistory.ts#L60 

How do you think the best way is to approach this if we don't want to have richHistory/addToRichHistory to have an override, want to pass all individual storage addToRichHistory calls through it for consistency and want to keep local storage's error about duplicate entries?

File: public/app/features/explore/state/query.ts [hunk 3]
```diff
@@ -486,11 +485,7 @@ async function handleHistory(
   queries: DataQuery[],
   exploreId: string
 ) {
-  const datasourceId = datasource.meta.id;
-  const nextHistory = updateHistory(history, datasourceId, queries);
-  dispatch(historyUpdatedAction({ exploreId, history: nextHistory }));
-
-  dispatch(addHistoryItem(datasource.uid, datasource.name, queries));
+  addHistoryItem(datasource.uid, datasource.name, queries);
```
Comment (ifrost) [inline public/app/features/explore/state/query.ts:488]: 🤔 addHistoryItem is still a thunk so should be dispatched, right?

```suggestion
  dispatch(addHistoryItem(datasource.uid, datasource.name, queries));
```

File: public/app/features/explore/state/history.ts [hunk 4]
```diff
@@ -79,20 +77,31 @@ export const addHistoryItem = (
...
+    // always add to local storage
+    getLocalRichHistoryStorage().addToRichHistory({
+      datasourceUid: datasourceUid,
+      datasourceName: datasourceName ?? '',
+      queries: queries,
+      starred: false,
+      comment: '',
+    });
+
+    if (config.queryHistoryEnabled) {
```
Comment (ifrost) [inline public/app/features/explore/state/history.ts:89]: Might be worth adding a comment why we do it only when `config.queryHistoryEnabled`.

File: public/app/features/explore/state/history.ts [hunk 5]
```diff
@@ -79,20 +77,31 @@ export const addHistoryItem = (
...
-      !getState().explore.richHistoryStorageFull,
-      !getState().explore.richHistoryLimitExceededWarningShown
-    );
-    if (richHistoryStorageFull) {
-      dispatch(richHistoryStorageFullAction());
-    }
-    if (limitExceeded) {
-      dispatch(richHistoryLimitExceededAction());
+    // always add to local storage
+    getLocalRichHistoryStorage().addToRichHistory({
```
Comment (ifrost) [inline public/app/features/explore/state/history.ts:81]: I think it'd be better to wrap it in `try/catch` as it may throw exceptions and prevent the remote history from being saved.

File: public/app/features/explore/state/history.ts [hunk 6]
```diff
@@ -79,20 +77,31 @@ export const addHistoryItem = (
...
-      '',
-      !getState().explore.richHistoryStorageFull,
-      !getState().explore.richHistoryLimitExceededWarningShown
-    );
-    if (richHistoryStorageFull) {
-      dispatch(richHistoryStorageFullAction());
-    }
-    if (limitExceeded) {
-      dispatch(richHistoryLimitExceededAction());
+    // always add to local storage
```
Comment (ifrost) [inline public/app/features/explore/state/history.ts:80]: The comment says exactly what the code, so maybe we can add a bit more info on "why"

```suggestion
    // always add to local storage - used to generate list of queries for autocomplete in Query Editors
```

File: public/app/features/explore/state/query.ts [hunk 7]
```diff
@@ -486,11 +485,11 @@ async function handleHistory(
   queries: DataQuery[],
   exploreId: string
```
Comment (ifrost) [inline public/app/features/explore/state/query.ts:486]: I think `history` and `exploreId` params are no longer used.

File: public/app/features/explore/state/query.ts [hunk 8]
```diff
@@ -486,11 +485,11 @@ async function handleHistory(
...
 ) {
-  const datasourceId = datasource.meta.id;
-  const nextHistory = updateHistory(history, datasourceId, queries);
-  dispatch(historyUpdatedAction({ exploreId, history: nextHistory }));
-
-  dispatch(addHistoryItem(datasource.uid, datasource.name, queries));
+  dispatch(addHistoryItem(true, datasource.uid, datasource.name, queries)); // always write to local for autocomplete, query history will use also if flag disabled
```
Comment (ifrost) [inline public/app/features/explore/state/query.ts:488]: Could it happen that:
- saving to local storage in line 488 fails (e.g. local storage is full)
- but then remote storage passes okay
- and the user see the error message for the local storage?
🤔 

It seems that for autocomplete we should ignore all the errors.

File: public/app/features/explore/state/utils.ts [hunk 9]
```diff
@@ -119,12 +123,55 @@ export async function loadAndInitDatasource(
     }
   }
 
-  const historyKey = `grafana.explore.history.${instance.meta?.id}`;
```
Comment (ifrost) [inline public/app/features/explore/state/utils.ts:122]: Does it mean that existing autocomplete history will not be accessible anymore? Wonder if we should keep reading from both for some time.

File: public/app/features/explore/state/query.ts (hunk 7 above)
Comment (gelicia) [reply public/app/features/explore/state/query.ts]: Great catch, thanks!

File: public/app/features/explore/state/query.ts (hunk 8 above)
Comment (gelicia) [reply public/app/features/explore/state/query.ts]: Okay, I implemented this with another parameter in `addHistoryItem` that will mute all errors and warnings. This does make the method signature to `addToRichHistory` a bit ridiculous, as there are are already booleans for showing the quota exceeded error and the limit exceeded warning. But, there is a generic dispatch for all errors that pop up (hiding the duplicate error) and I thought it made sense to make that a separate "permission" of sorts. Let me know if you disagree.

File: public/app/features/explore/state/utils.ts (hunk 9 above)
Comment (gelicia) [reply public/app/features/explore/state/utils.ts:122]: Added back! The history now will add in query history, query history of mixed queries where the individual query matches the datasource, then local storage query history entries. It will still stop trying to add if we ever hit the max number, and will cap the final result to the max number.

File: public/app/features/explore/state/query.ts (hunk 8 above)
Comment (ifrost) [reply public/app/features/explore/state/query.ts]: > This does make the method signature to `addToRichHistory` a bit ridiculous,

Would creating and using a wrapper for the remote storage make it more readable? 🤔  Something like:

```typescript
export default class DualWriteRichHistoryRemoteStorage extends RichHistoryRemoteStorage {

  autocompleteHistory: RichHistoryLocalStorage;
  
  constructor(props) {
    super(props);
    this.autocompleteHistory = new RichHistoryLocalStorage();
  }

  async addToRichHistory(
    newRichHistoryQuery: Omit<RichHistoryQuery, 'id' | 'createdAt'>
  ): Promise<{ warning?: RichHistoryStorageWarningDetails; richHistoryQuery: RichHistoryQuery }> {
    const result = await super.addToRichHistory(newRichHistoryQuery);
    try {
      this.autocompleteHistory.addToRichHistory(newRichHistoryQuery);
    } catch(e) {}
    return result;
  }

}
```

and then create DualWriteRichHistoryRemoteStorage when flag is enabled:

```typescript
const richHistoryLocalStorage = new RichHistoryLocalStorage();
const richHistoryRemoteStorage = new DualWriteRichHistoryRemoteStorage();

export const getRichHistoryStorage = (): RichHistoryStorage => {
  return config.queryHistoryEnabled ? richHistoryRemoteStorage : richHistoryLocalStorage;
};
```

Comment (gelicia) [reply public/app/features/explore/state/query.ts]: @ifrost the wrapper idea is a good one, but I just think another layer in all this makes it hard to follow. Can you check out what I have now? I made `richHistory/addToRichHistory` take an object with some sensible defaults, and I think we can make calling it more clean now without as much wrapper-ing.

Comment (ifrost) [reply public/app/features/explore/state/query.ts]: Looks good at first glance 👍  I'll try to test it locally tomorrow. Do you think renaming actions a bit would make sense to reflect what they mean now and have:

```
richHistoryStorageFullErrorShownAction
richHistoryLimitExceededWarningShownAction
```
?

File: public/app/features/explore/state/utils.ts [hunk 10]
```diff
@@ -119,12 +123,64 @@ export async function loadAndInitDatasource(
...
-  const history = store.getObject<HistoryItem[]>(historyKey, []);
-  // Save last-used datasource
+  let history: HistoryItem[] = [];
+
+  const localStorageHistory = getLocalRichHistoryStorage();
+
+  const historyResults = await localStorageHistory.getRichHistory({
+    search: '',
+    sortOrder: SortOrder.Ascending,
+    datasourceFilters: [instance.name],
```
Comment (ifrost) [inline public/app/features/explore/state/utils.ts:133]: When mixed is selected as the root, could we filter by multiple data sources in here? 🤔  At the moment I don't see history for a data source that is used as a child of mixed (in the video: when switched to mixed I still could see "created in non-mixed mode" entry):


https://github.com/grafana/grafana/assets/745532/2c19f8b8-6193-4e42-8549-21c857a58c6c

Comment (gelicia) [reply public/app/features/explore/state/utils.ts:133]: @ifrost Its not impossible but I'd like to make that a separate piece of work due to its complexity and the necessity of changing a core component.

The logic to bring history into queries is [loadAndInitDatasource](https://github.com/grafana/grafana/blob/main/public/app/features/explore/state/utils.ts#L99) which only brings in a single datasource, and returns only that datasource's history. We could have places that call this function also pass in the datasources available in the queries, and then return those histories as well. What do you think about a ticket doing the following:

Change loadAndInitDatasource to have as parameters
- orgId: number,
- rootDatasource: DataSourceRef | string
- queryDatasources? DataSourceRef | string[]

and return `Promise<{ queryHistories: Array<{ refId: string; history: HistoryItem[]}>; rootInstance: DataSourceApi }>`

With this change, if (for example) the root datasource is mixed with loki as the datasource for query refId A, it will be called with `(1, mixed, [loki])` and return `[{refId: 'A', history: [loki history]}]` as the history. 

We will need to modify how history is kept in state, as it is now just assuming one history for query https://github.com/grafana/grafana/blob/main/public/app/types/explore.ts#L126

I think the structure `Array<{ refId: string; history: HistoryItem[]}>` would work fine for state as well.

Finally, explore now retrieves one history for every row ([src](https://github.com/grafana/grafana/blob/main/public/app/features/explore/QueryRows.tsx#L44)) and passes it onto the core component `QueryEditorRows`. `QueryEditorRows` attaches the same history to every query ([src](https://github.com/grafana/grafana/blob/main/public/app/features/query/components/QueryEditorRows.tsx#L179)). But we could make it attach the right history for the given RefId - as long as we're sure to handle the history change if the query datasource changes.

Let me know if there's something I'm missing, if you'd like me to tackle this in this PR, or if I should write up an issue. I can get started on that right away.

Comment (gelicia) [reply public/app/features/explore/state/utils.ts:133]: Oh, a drawback to this is that there will be some redundancy - like if RefId A and B use the same loki datasource, it would have the history for both seperately. It is more straightforward than querying off datasource UID but that would be more efficient. Let me know if youd rather I do that.

File: public/app/features/explore/state/utils.ts [hunk 11]
```diff
@@ -119,12 +123,64 @@ export async function loadAndInitDatasource(
...
+  }
+
+  if (history.length < MAX_HISTORY_AUTOCOMPLETE_ITEMS) {
+    // check the last 100 mixed history results seperately
+    const historyMixedResults = await localStorageHistory.getRichHistory({
+      search: '',
+      sortOrder: SortOrder.Ascending,
+      datasourceFilters: [MIXED_DATASOURCE_NAME],
+      from: 0,
+      to: MAX_HISTORY_AUTOCOMPLETE_ITEMS,
```
Comment (ifrost) [inline public/app/features/explore/state/utils.ts:155]: I think it's not well documented but `from` / `to` is used as a time filter where value is equal to "number of days since today" . `to: MAX_HISTORY_AUTOCOMPLETE_ITEMS` means getting queries that are at max 100 days old. Is `MAX_HISTORY_AUTOCOMPLETE_ITEMS` used just to have some arbitrary big number? If so I think it be more readable if it was a separate const or if we changed the API to return all items.

File: public/app/features/explore/state/utils.ts (hunk 10 above)
Comment (ifrost) [reply public/app/features/explore/state/utils.ts:133]: Oh I thought it was as simple as changing `datasourceFilters: [instance.name],` to use names from each query but if I understand correctly we would need to load the data source to get it's name. 

The current implementation makes things better anyway, so I'm happy to keep  not include it in this PR.

File: public/app/features/explore/state/utils.ts (hunk 11 above)
Comment (gelicia) [reply public/app/features/explore/state/utils.ts]: So the to/from date filter is not required by the API, so I made it optional, then dont pass it in for local storage. The only place where it needs a time filter is the queries tab, which already has the default of going from now to the retention period of 2 weeks ([src](https://github.com/grafana/grafana/blob/main/public/app/features/explore/RichHistory/RichHistoryQueriesTab.tsx#L143)).

File: public/app/features/explore/state/utils.ts [hunk 12]
```diff
@@ -119,12 +123,60 @@ export async function loadAndInitDatasource(
...
+        historyResult.queries.forEach((q) => {
+          if (q?.datasource?.uid === instance.uid) {
+            history.push({ ts: parseInt(historyResult.id, 10), query: q });
+          }
+        });
+      });
+    }
+  }
+
+  // finally, add any legacy local storage history that might exist. To be removed in Grafana 11
```
Comment (ifrost) [inline public/app/features/explore/state/utils.ts:167]: Do we have a task for it in the backlog?

**Review by ifrost — APPROVED** [review]
